compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -963,8 +963,10 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
 
     /** `tree.isInstanceOf[tp]`, with special treatment of singleton types */
     def isInstance(tp: Type)(using Context): Tree = tp.dealias match {
+      case ConstantType(c) if c.tag == StringTag =>
```

Comment:
Shouldn't we do it for other constants?

Reply:
The `String` only constant types supporting the alternative comparison with `eq` are classes and strings. Class literals are hash consed, so `==` and `eq` are the same. So I believe `String` is the only type in need of a fix.

Reply:
To add, I didn't find this issue with integer constants.

Reply:
`Null` might also take advantage of this unless some other optimization is already handling it.